### PR TITLE
`sc-block-builder`: Remove `BlockBuilderProvider`

### PR DESCRIPTION
**Pull request by bkchr**
The `BlockBuilderProvider` was a trait that was defined in `sc-block-builder`. The trait was implemented for `Client`. This basically meant that you needed to import `sc-block-builder` any way to have access to the block builder. So, this trait was not providing any real value. This pull request is removing the said trait. Instead of the trait it introduces a builder for creating a `BlockBuilder`. The builder currently has the quite fabulous name `BlockBuilderBuilder` (I'm open to any better name :sweat_smile:). The rest of the pull request is about replacing the old trait with the new builder.

# Downstream code changes

If you used `new_block` or `new_block_at` before you now need to switch it over to the new `BlockBuilderBuilder` pattern:

```rust
// `new` requires a type that implements `CallApiAt`. 
let mut block_builder = BlockBuilderBuilder::new(client)
                // Then you need to specify the hash of the parent block the block will be build on top of
		.on_parent_block(at)
                // The block builder also needs the block number of the parent block. 
                // Here it is fetched from the given `client` using the `HeaderBackend`
                // However, there also exists `with_parent_block_number` for directly passing the number
		.fetch_parent_block_number(client)
		.unwrap()
                // Enable proof recording if required. This call is optional.
		.enable_proof_recording()
                // Pass the digests. This call is optional.
                .with_inherent_digests(digests)
		.build()
		.expect("Creates new block builder");
```

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4143248

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4143305

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4143307

File: substrate/client/block-builder/src/lib.rs [hunk 1]
```diff
@@ -29,56 +29,147 @@
...
+			call_api_at: self.call_api_at,
+			enable_proof_recording: false,
+			inherent_digests: Default::default(),
+			parent_block: self.parent_block,
+			parent_number,
 		}
 	}
 }
 
+/// The second stage of the [`BlockBuilderBuilder`].
```
Comment (dmitry-markin) [inline substrate/client/block-builder/src/lib.rs:126]:
```suggestion
/// The third stage of the [`BlockBuilderBuilder`].
```
Nitpick: may be rename `Stage1`->`Stage2`, `Stage2`->`Stage3`?

File: substrate/primitives/api/test/tests/runtime_calls.rs [hunk 2]
```diff
@@ -105,9 +105,12 @@ fn record_proof_works() {
 	.into_unchecked_extrinsic();
 
 	// Build the block and record proof
-	let mut builder = client
-		.new_block_at(client.chain_info().best_hash, Default::default(), true)
-		.expect("Creates block builder");
+	let mut builder = BlockBuilderBuilder::new(&client)
+		.on_parent_block(&client, client.chain_info().best_hash)
```
Comment (dmitry-markin) [inline substrate/primitives/api/test/tests/runtime_calls.rs:109]: Why do we pass `&client` to `on_parent_block` while it's already passed to `new` and `on_parent_block` takes only one argument (hash)?

**Review by dmitry-markin — APPROVED** [review]

File: substrate/primitives/api/test/tests/runtime_calls.rs (hunk 2 above)
Comment (bkchr) [reply substrate/primitives/api/test/tests/runtime_calls.rs:109]: An oversight :D I had refactored it in the lifetime of the pr :D

**Review by skunert — APPROVED** [review]

**Review by skunert — APPROVED** [review]
Logic looks good! Some missed conversions to the new builder, but CI notified you of that already :D

**Review by skunert — APPROVED** [review]
Logic looks good! Some missed conversions to the new builder, but CI notified you of that already :D

File: substrate/client/basic-authorship/src/basic_authorship.rs [hunk 3]
```diff
@@ -335,8 +310,12 @@ where
 		block_size_limit: Option<usize>,
 	) -> Result<Proposal<Block, PR::Proof>, sp_blockchain::Error> {
 		let block_timer = time::Instant::now();
-		let mut block_builder =
-			self.client.new_block_at(self.parent_hash, inherent_digests, PR::ENABLED)?;
+		let mut block_builder = BlockBuilderBuilder::new(&*self.client)
+			.on_parent_block(self.parent_hash)
+			.fetch_parent_block_number(&*self.client)?
```
Comment (michalkucharczyk) [inline substrate/client/basic-authorship/src/basic_authorship.rs:315]: `self.parent_number` could be used, as it is already known.

File: substrate/client/block-builder/src/lib.rs (hunk 1 above)
Comment (michalkucharczyk) [reply substrate/client/block-builder/src/lib.rs:126]: or maybe builder could be enum? Not sure however if that would work.

File: substrate/client/block-builder/src/lib.rs [hunk 4]
```diff
@@ -29,56 +29,147 @@
...
+where
+	B: BlockT,
+{
+	/// Fetch the parent block number from the given `header_backend`.
+	///
+	/// The parent block number is used to initialize the block number of the new block.
+	///
+	/// Returns an error if there doesn't exist a number for the parent block specified in
+	/// [`on_parent_block`](BlockBuilderBuilder::on_parent_block).
+	pub fn fetch_parent_block_number<H: HeaderBackend<B>>(
```
Comment (michalkucharczyk) [inline substrate/client/block-builder/src/lib.rs:88]: nit: not sure if we really need this function. number could be fetched outside.

Comment (michalkucharczyk) [reply substrate/client/block-builder/src/lib.rs:88]: Or maybe we could provide variant with parent header?

**Review by michalkucharczyk — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4174576

File: substrate/client/block-builder/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply substrate/client/block-builder/src/lib.rs:126]: No enum would not work. The idea why these are multiple different structs is to require that the user is configuring required things.

Comment (bkchr) [reply substrate/client/block-builder/src/lib.rs:126]: With an enum that would not work.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4185906

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4187805

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4200421

**Comment by bkchr** [comment]
bot fmt

**Pull request merged**